### PR TITLE
NO-ISSUE: oauth-apiserver: apply GenericServerRunOptions to the server configuration

### PR DESCRIPTION
**Pull request by p0lyn0mial**
Before, some flags like `ShutdownDelayDuration` were never applied to the server.
Wthout this path, [these](https://github.com/openshift/oauth-apiserver/blob/99607e9c37db24fc6a864c38a175f45383e0a4a1/vendor/k8s.io/apiserver/pkg/server/options/server_run_options.go#L113) options weren't applied.

Without `ShutdownDelayDuration` the server can be immediately terminated when all in-flight requests are completed.

Should improve https://sippy.dptools.openshift.org/sippy-ng/component_readiness/test_details?arch=amd64&baseEndTime=2024-02-28%2023%3A59%3A59&baseRelease=4.15&baseStartTime=2024-02-01%2000%3A00%3A00&capability=operator-conditions&component=oauth-apiserver&confidence=95&environment=ovn%20no-upgrade%20amd64%20metal-ipi%20standard&excludeArches=arm64%2Cheterogeneous%2Cppc64le%2Cs390x&excludeClouds=openstack%2Cibmcloud%2Clibvirt%2Covirt%2Cunknown&excludeVariants=hypershift%2Cosd%2Cmicroshift%2Ctechpreview%2Csingle-node%2Cassisted%2Ccompact&groupBy=cloud%2Carch%2Cnetwork&ignoreDisruption=true&ignoreMissing=false&minFail=3&network=ovn&pity=5&platform=metal-ipi&sampleEndTime=2024-03-05%2023%3A59%3A59&sampleRelease=4.16&sampleStartTime=2024-02-28%2000%3A00%3A00&testId=Operator%20results%3A466e1a49a33b63218495dc8201953194&testName=operator%20conditions%20authentication&upgrade=no-upgrade&variant=standard

Should be backported to previous releases.

UPDATE:

I have analysed [another CI ](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_oauth-apiserver/105/pull-ci-openshift-oauth-apiserver-master-e2e-aws/1765655024194555904)run today and here is what I have found so far:

In general, in this run, the `OpenShiftAPICheckFailed` event for `oauth.openshift.io.v1` was reported 94 times and for `user.openshift.io.v1` 95 times.

`T 08:51:14`  -  `kubelet` reports an oauth-apiserver instance as ready `SyncLoop (probe)" probe="readiness" status="ready" pod="openshift-oauth-apiserver/apiserver-58f894bf9-zt7tf` . This was the first instance. The operator almost immediately picked it up and tried to connect to that instance.

`T 08:51:14`    - the `authentication-operato`r reports `Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-authentication-operator", Name:"authentication-operator", UID:"5a175510-fdce-410d-8250-f3ac80b12fd0", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'OpenShiftAPICheckFailed' "oauth.openshift.io.v1" failed with an attempt failed with statusCode = 503, err = the server is currently unable to handle the request` This was the first entry. Who was that instance talking to, given that the first kas instance was installed 1m15s later?

`T: 08:52:33` - `kube-apiserver-ip-10-0-17-232.ec2.interna`l instance is installed. This was the first instance, the remaining instances were installed later.

`T: 08:52:35` - the `available_controller` starts flapping `changing APIService availability" name="v1.oauth.openshift.io" oldStatus="False" newStatus="True" message="all checks passed" reason="Passed`
The flapping indicates that the instance had a fight with another instance. What is interesting here is that I didn’t find entries changing the status from True to False . Who was that instance fighting with when the other instances weren’t even installed?

`T 08:54:22` - Destroying the bootstrap resources - (`openshift_install-1709802470.log`)

`T 08:54:27` - the `authentication-operator`   reports `Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-authentication-operator", Name:"authentication-operator", UID:"5a175510-fdce-410d-8250-f3ac80b12fd0", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'OpenShiftAPICheckFailed' "user.openshift.io.v1" failed with an attempt failed with statusCode = 503, err = the server is currently unable to handle the request` for the last time.

Hypothesis: it looks like the bootstrap API server caused the outage that was reported by the operator.

### CODE REVIEW

File: pkg/cmd/oauth-apiserver/cmd_test.go [hunk 1]
```diff
@@ -39,3 +54,148 @@ func TestServerStartUp(t *testing.T) {
...
+	expected := &oauthapiserver.OAuthAPIServerOptions{
+		GenericServerRunOptions: &genericapiserveroptions.ServerRunOptions{
+			MaxRequestsInFlight:         400,
+			MaxMutatingRequestsInFlight: 200,
+			RequestTimeout:              time.Minute,
+			MinRequestTimeout:           1800,
+			ShutdownDelayDuration:       time.Second * 7,
+			JSONPatchMaxCopyBytes:       int64(3 * 1024 * 1024),
+			MaxRequestBodyBytes:         int64(3 * 1024 * 1024),
+			EnablePriorityAndFairness:   true,
```
Comment (p0lyn0mial) [inline pkg/cmd/oauth-apiserver/cmd_test.go:81]: I will address this in a separate PR. I think that we should disable PnF for the aggregated servers.
@tkashem wdyt ?

Comment (p0lyn0mial) [reply pkg/cmd/oauth-apiserver/cmd_test.go:81]: This is preexisting - that is PnF is enabled without this PR.

Comment (benluddy) [reply pkg/cmd/oauth-apiserver/cmd_test.go:81]: You might save effort if you do this after rebasing to 1.29: https://github.com/kubernetes/kubernetes/pull/118886.

Comment (p0lyn0mial) [reply pkg/cmd/oauth-apiserver/cmd_test.go:81]: sure but how are we going to fix the previous versions ?

Comment (p0lyn0mial) [reply pkg/cmd/oauth-apiserver/cmd_test.go:81]: I will try to backport it to 1.28

Comment (p0lyn0mial) [reply pkg/cmd/oauth-apiserver/cmd_test.go:81]: @benluddy thanks for pointing out the patch, it looks like it actually works!

Comment (tkashem) [reply pkg/cmd/oauth-apiserver/cmd_test.go:81]: yes, I agree, APF should be disabled

File: pkg/cmd/oauth-apiserver/cmd_test.go [hunk 2]
```diff
@@ -39,3 +54,148 @@ func TestServerStartUp(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestAddFlags(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/cmd/oauth-apiserver/cmd_test.go:60]: decided to add a test (`TestOAuthAPIServerConfig`) to this PR

**Comment by p0lyn0mial** [comment]
/assign @stlaz

for approval

File: pkg/cmd/oauth-apiserver/cmd.go [hunk 3]
```diff
@@ -127,6 +127,11 @@ func (o *OAuthAPIServerOptions) NewOAuthAPIServerConfig() (*apiserver.Config, er
 	}
 
 	serverConfig := apiserver.NewConfig()
+
+	if err := o.GenericServerRunOptions.ApplyTo(&serverConfig.GenericConfig.Config); err != nil {
```
Comment (stlaz) [inline pkg/cmd/oauth-apiserver/cmd.go:131]: would `github.com/openshift/library-go/pkg/config/configdefaults.SetRecommendedGenericAPIServerConfigDefaults` work?

Comment (p0lyn0mial) [reply pkg/cmd/oauth-apiserver/cmd.go:131]: no, `GenericServerRunOptions.ApplyTo` is applying options - get from the command lines arguments to a configuration. We should respect command line options set by the operator or administrator.

Comment (stlaz) [reply pkg/cmd/oauth-apiserver/cmd.go:131]: oh, that's a nasty bug we've got here then :eyes:

**Pull request merged**